sklearn/linear_model/_glm/tests/test_glm.py

```diff
@@ -0,0 +1,355 @@
+                                     link='identity', fit_intercept=True,
+                                     solver=solver, check_input=False,
+                                     max_iter=300)
+    glm.fit(X_train, y_train)
+    assert glm.coef_.shape == (X.shape[1], )
+    assert_allclose(glm.coef_, ridge.coef_, atol=5e-5)
+    assert_allclose(glm.intercept_, ridge.intercept_, rtol=1e-5)
+    assert_allclose(glm.predict(X_train), ridge.predict(X_train), rtol=5e-5)
+    assert_allclose(glm.predict(X_test), ridge.predict(X_test), rtol=5e-5)
+
+
+@pytest.mark.parametrize('solver, tol', [('lbfgs', 1e-7)])
+def test_poisson_glmnet(solver, tol):
+    """Compare Poisson regression with L2 regularization and LogLink to glmnet
+    """
+    # library("glmnet")
+    # options(digits=10)
+    # df <- data.frame(a=c(-2,-1,1,2), b=c(0,0,1,1), y=c(0,1,1,2))
+    # x <- data.matrix(df[,c("a", "b")])
+    # y <- df$y
+    # fit <- glmnet(x=x, y=y, alpha=0, intercept=T, family="poisson",
+    #               standardize=F, thresh=1e-10, nlambda=10000)
+    # coef(fit, s=1)
+    # (Intercept) -0.12889386979
+    # a            0.29019207995
+    # b            0.03741173122
+    X = np.array([[-2, -1, 1, 2], [0, 0, 1, 1]]).T
+    y = np.array([0, 1, 1, 2])
+    glm = GeneralizedLinearRegressor(alpha=1,
+                                     fit_intercept=True, family='poisson',
+                                     link='log', tol=1e-7,
+                                     solver=solver, max_iter=300,
+                                     )
+    glm.fit(X, y)
+    assert_allclose(glm.intercept_, -0.12889386979, rtol=1e-5)
+    assert_allclose(glm.coef_, [0.29019207995, 0.03741173122], rtol=1e-5)
+
+
+@pytest.mark.parametrize("solver", GLM_SOLVERS)
+def test_convergence_warning(solver, regression_data):
+    X, y = regression_data
+
+    est = GeneralizedLinearRegressor(solver=solver,
+                                     max_iter=1, tol=1e-20)
+    with pytest.warns(ConvergenceWarning):
+        est.fit(X, y)
+
+
+def test_poisson_regression_family(regression_data):
+    est = PoissonRegressor()
+    est.family == "poisson"
+
+    msg = "PoissonRegressor.family must be 'poisson'!"
+    with pytest.raises(ValueError, match=msg):
+        est.family = 0
+
+
+def test_gamma_regression_family(regression_data):
+    est = GammaRegressor()
+    est.family == "gamma"
+
+    msg = "GammaRegressor.family must be 'gamma'!"
+    with pytest.raises(ValueError, match=msg):
+        est.family = 0
+
+
+def test_tweedie_regression_family(regression_data):
+    power = 2.0
+    est = TweedieRegressor(power=power)
+    assert isinstance(est.family, TweedieDistribution)
+    assert est.family.power == power
+    msg = "TweedieRegressor.family must be of type TweedieDistribution!"
+    with pytest.raises(TypeError, match=msg):
+        est.family = None
+
+    # TODO: the following should not be allowed
+    # est.family.power = 2
```

Comment:
cleanup

Reply:
I reformulated the comment a bit. I think this comment can still be useful, but I don't intend to address it as all the solutions I looked are somewhat convoluted.

Briefly the issue is that if one changes `TweedieRegressor.power` all is well and `TweedieRegressor.family.power` will be updated accordingly. If one tried to set the `TweedieRegressor.family` an error will be raised as expected (see notes in `glm.py:L653` for mode details). However, setting `TweedieRegressor.family.power` will not update `TweedieRegressor.power` leading to `get_params` inconsistent with the params actually used.

It shouldn't be an issue unless user changes `TweedieRegressor.family.power` intentionally in which case he should know what he is doing. If we want to disallow that, we would need to implement a read-only attrs in `ExponentialDispersionModel`, e.g. by defining custom `__setattr__` with an init parameter, or re-thinking this whole inheritance mechanism for `GeneralizedLinearRegressor` completely.

I'm not sure it's really a issue that needs solving though given the benefit/complexity cost.

Reply:
ok then. Can you make this code a proper test of an error or just remove it?

Reply:
I can't make it test for an error, because this doesn't error now. The comment meant to say that it might be good to error in the future. Anyway removed it.